---
title: ""
output: 
  md_document
---

```{r, echo = FALSE, message = FALSE, warning = FALSE}
#knitr::opts_chunk$set(out.width='750px', dpi=200)
knitr::opts_chunk$set(collapse = TRUE, comment = "#>")
```


[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/insect)](https://cran.r-project.org/package=insect)
[![CRAN_Downloads_Badge](http://cranlogs.r-pkg.org/badges/grand-total/insect)](https://cran.r-project.org/package=insect)
[![DOI](https://zenodo.org/badge/87808693.svg)](https://zenodo.org/badge/latestdoi/87808693)
[![ORCiD](https://img.shields.io/badge/ORCiD-0000--0002--7332--7931-brightgreen.svg)](http://orcid.org/0000-0002-7332-7931) 
[![Build_Status](https://travis-ci.org/shaunpwilkinson/insect.svg?branch=master)](https://travis-ci.org/shaunpwilkinson/insect)
[![Project_Status](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)
[![License](https://img.shields.io/badge/License-GPL%20v3-blue.svg)](http://www.gnu.org/licenses/gpl-3.0)

--------------------------------------------------------------------------------

### Informatic sequence classification trees

`insect` is an R package for taxonomic identification of amplicon 
sequence variants generated by DNA meta-barcoding analysis. 
The learning and classification algorithms implemented in the 
package are based on full probabilistic models (profile hidden Markov models) 
and offer highly accurate taxon IDs, albeit at a relatively high computational cost.

The package also contains functions for searching and downloading reference 
sequences and taxonomic information from NCBI, 
a "virtual PCR" tool for sequence trimming, 
a function for "purging" erroneously labeled reference sequences, 
and several other handy tools.  

`insect` is designed to be used in conjunction with the 
[dada2](https://benjjneb.github.io/dada2/index.html) pipeline or any other
de-noising tool that produces a list of amplicon sequence variants (ASVs). 
While unfiltered sequences can also be processed with high accuracy, 
the **insect** classification algorithm is relatively slow, 
since it uses a computationally intensive dynamic
programming algorithm to find the likelihood values
of each sequence given the models at each node of the classification tree. 
Hence an appropriately filtered input dataset will generally be 
much faster to process.


### Installation

To download **insect** from CRAN and load the package, run

```{r, eval = FALSE}
install.packages("insect")
library(insect)
```

To download the latest development version from GitHub, run:

```{r, eval = FALSE}
devtools::install_github("shaunpwilkinson/insect", build_vignettes = TRUE) 
library(insect)
```

```{r, echo = FALSE}
library(insect)
```



### Classifying sequences

Training an **insect** classifier will be the subject of another tutorial;
however, several classifiers are already available for some of the more common metabarcoding primer
sets:

<!-- note newlines needed between html tags and code chunk -->

<div class="verysmall">

```{r, echo = FALSE, results='asis'}
tmp <- tempfile(fileext = ".csv")
u <- "https://docs.google.com/spreadsheets/d/1pmTlZBnWIZzxZzS8zm943uGL56qpDPBmoOBht9OzVvQ/export?gid=0&format=csv"
download.file(u, destfile = tmp)
mytab <- read.table(text = readLines(tmp, warn = FALSE), header = TRUE,  sep = ",", stringsAsFactors = FALSE)
mytab <- mytab[order(mytab$Marker, mytab$Target),]
rownames(mytab) <- NULL
knitr::kable(mytab)
```

</div>

To classify a sequence or set of sequences, first read them into R as a "DNAbin" 
list object. FASTA files can be parsed as follows:

```{r, eval = FALSE}
x <- readFASTA("<path-to-file>.fasta")
```

Alternatively users may wish to assign taxon IDs to the output from the 
[DADA2](https://www.nature.com/articles/nmeth.3869)
pipeline, in which case the column names of the ouput
table can be parsed as in the following example:

```{r}
data("samoa") 
x <- char2dna(colnames(samoa))
```

The next step is to download and read in the classifier.
It is important to ensure that the classifier was trained using the
same primer set as that used to generate the query data.
In this example the data were generated from 
autonomous reef monitoring structures in 
American Samoa (ARMS) using the COI metabarcoding primers mlCOIintF and jgHCO2198 
([Leray et al 2013](https://frontiersinzoology.biomedcentral.com/articles/10.1186/1742-9994-10-34)),
and de-noised, filtered and merged following the 
[DADA2 tutorial](https://benjjneb.github.io/dada2/tutorial.html).

There are currently three classifiers available for this primer set, 
a [marine only](https://www.dropbox.com/s/vn1yit2wkug9f7p/classifier.rds?dl=1) version,
one for DNA sequences with every third codon position removed 
([nothirds](https://www.dropbox.com/s/bfvqnggi4jt591i/classifier.rds?dl=1)), 
and one for translated amino acid sequences 
([amino](https://www.dropbox.com/s/c987wohded37cxg/classifier.rds?dl=1)).

Here we'll use the amino acid version, which was created by translating the 
[MIDORI UNIQUE 20180221](http://reference-midori.info/download.php)
training dataset using the [EBI5](https://www.ebi.ac.uk/ena/browse/translation-tables) 
invertebrate mitochondrial translation table.

The 47 MB classifier can be downloaded to the current working directory 
and read into R as follows:

```{r, eval = FALSE}
download.file("https://www.dropbox.com/s/c987wohded37cxg/classifier.rds?dl=1", 
              destfile = "classifier.rds", mode = "wb")
classifier <- readRDS("classifier.rds")
```

```{r, echo = FALSE}
classifier <- readRDS("~/Dropbox/R/insect/trees/COI/metazoan_COI/version_4/amino/classifier.rds")
```

Now we need to translate the amplicon sequence variants (ASVs) 
to amino acid sequences using the same number 5 translation table. 
Before this, any sequences
whose length is not divisible by three should be discarded. 
The sequences are then converted to character type for translation 
with the `translate` function in the 
[seqinr](https://cran.r-project.org/package=seqinr) package, and
converted back to binary form object using `ape::as.AAbin`.

```{r}
keeps <- sapply(x, length) %% 3 == 1L
x <- x[keeps]
x <- ape::as.character.DNAbin(x)
x <- lapply(x, seqinr::translate, numcode = 5, frame = 1)
x <- ape::as.AAbin(x)
```

Now we need to check that none of the translated sequences contain stop codons:

```{r}
keeps <- sapply(x, function(v) !any(v == as.raw(42)))
x <- x[keeps]
```

Finally, the amino acid sequences are passed to the classifier for taxon ID
assignment.
There is an option to perform a nearest-neighbor search prior to the 
computationally-expensive recursive model test procedure, which can save 
time and improve resolution ('recall') at lower taxonomic ranks.
Note that this can be a double-edged sword; if multiple species share
an identical or near-identical sequence, and the true taxon of the query sequence
is missing from the trainingset, the algorithm may over-classify the sequence
and return a congeneric taxon.
To perform a nearest-neighbor search with a similarity threshold of say 0.99
(meaning any sequence in the trainingset with a similarity greater than or 
equal to 0.99 is considered a match), set `ping = 0.99`.


```{r}
out <- classify(x, classifier, cores = 2, ping = 0.99)
```


<!-- note newlines needed between html tags and code chunk -->

<div class="verysmall">

```{r, echo = FALSE}
knitr::kable(out)
```

</div>


Any sequences that return exact hits or near matches (> 99% similarity in this case) 
with at least one training sequence are assigned 
a score of `NA`, as in the final row of the table above. 
Here, the multiple matching sequences have a Nereid polychaete common ancestor,
and the query sequence was therefore assigned to the family Nereidae.


### Further reading

A more detailed overview of the package and its functions can be found
[here](https://rpubs.com/shaunpwilkinson/insect) or by running

```{r, eval = FALSE}
vignette("insect-vignette")
```

### Issues

If you experience a problem using this software please feel free to
raise it as an issue on [GitHub](http://github.com/shaunpwilkinson/insect/issues).


### Acknowledgements

This software was developed at 
[Victoria University of Wellington](http://www.victoria.ac.nz/) 
with funding from a Rutherford Foundation Postdoctoral Research Fellowship 
award from the Royal Society of New Zealand.


